### PR TITLE
Add lost qualifier to viewpointtest type checker

### LINKED ISSUE #576: Lost qualifier should be added to type systems with viewpoint adaptation functionality.
Imagine a standard lattice like this:
```
      TOP
       |
       |
       v
A      B       RDQ
       |
       |
       v
     BOTTOM
```
where TOP is the top type in the lattice, BOTTOM is the bottom type in the lattice and RDQ (receiver-dependent-qualifier) is for viewpoint adaptation.

Suppose this type system also supports viewpoint adaptation and a piece of code written like this:
```java
   class Demo {
     @RDQ Object f;

    void adaptation(@Top Demo top) {
        top.f = new @A Object(); // allowed
    }
   }
```
This looks fine, top.f has type `TOP |> RDQ  = TOP` by the viewpoint adaptation law and thus can be assigned by a subtype object (the @A Object)

However, if the method is invoked like this, we would lost our type safety.
```java
@B Demo o;
adaptation(o);
```
In the above code `o` has type `B` and `o.f` has type `B` as well but in the method invocation, the field is assigned to a type `A` object which is not a subtype of type `B`.

This problem could cause serious type safety issues in some type systems, for example, the immutability type system.
We could add a lost qualifier (which was introduced in the Generic Universe Type paper) as the type when the top qualifier looking at a RDQ qualifier. In that case if the variable has type lost, assignment operation is forbidden.

### PR DESCRIPTION
**Pull request by Ao-senXiong**
Fixes #576

### CODE REVIEW

**Comment by Ao-senXiong** [comment]
> Fixes #576

TODO: lost is not reflexive

**Comment by Ao-senXiong** [comment]
> > Fixes #576
> 
> TODO: lost is not reflexive

This is implemented in [76a9f40](https://github.com/eisop/checker-framework/pull/827/commits/76a9f403cf5d439f505649090efc9fe26eaebc8a), but how to improve the error message remain as a questions.

File: framework/src/test/java/viewpointtest/ViewpointTestQualifierHierarchy.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import java.lang.annotation.Annotation;
+import java.util.Collection;
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.util.Elements;
+
+import viewpointtest.quals.Bottom;
+import viewpointtest.quals.Lost;
+
+/** A qualifier hierarchy for the ViewpointTest checker. */
```
Comment (wmdietl) [inline framework/src/test/java/viewpointtest/ViewpointTestQualifierHierarchy.java:15]:
```suggestion
/** A qualifier hierarchy for the Viewpoint Test Checker. */
```

Or a link to the class?

File: framework/src/test/java/viewpointtest/quals/Lost.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE_USE, ElementType.TYPE_PARAMETER})
+@SubtypeOf({Top.class})
+public @interface Lost {}
```
Comment (wmdietl) [inline framework/src/test/java/viewpointtest/quals/Lost.java:15]: It is interesting that the requirement of javadoc doesn't fail on this file (and also on `Bottom.java` above)... it that check not enabled for `framework/src/test/java/`?
Can you make a separate PR before this one to ensure test sources are also checked?

File: framework/tests/viewpointtest/LostNonReflexive.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import viewpointtest.quals.*;
+
+public class LostNonReflexive {
+    @ReceiverDependentQual Object f;
+
+    @SuppressWarnings({"inconsistent.constructor.type", "super.invocation.invalid"})
+    @ReceiverDependentQual
+    LostNonReflexive(@ReceiverDependentQual Object args) {}
+
+    @ReceiverDependentQual
```
Comment (wmdietl) [inline framework/tests/viewpointtest/LostNonReflexive.java:10]: In a separate PR, can you ensure these test qualifiers are added to type annotation formatting: https://github.com/eisop/checker-framework/blob/930212447c1a39bc9cdfa82b4b0b690a9739b27b/build.gradle#L303C36-L303C73

Maybe there is some external way to set this in Spotless? Or some other way to do this more nicely.

File: framework/tests/viewpointtest/VarargsConstructor.java [hunk 4]
```diff
@@ -17,6 +17,7 @@ void foo() {
     void invokeConstructor(@A Object aObj, @B Object bObj, @Top Object topObj) {
         @A Object a = new @A VarargsConstructor(aObj);
         @B Object b = new @B VarargsConstructor(bObj);
+        // :: error: (argument.type.incompatible)
         @Top Object top = new @Top VarargsConstructor(topObj);
```
Comment (wmdietl) [inline framework/tests/viewpointtest/VarargsConstructor.java:21]: I think this is a tricky one.
For other uses of Top, we want Lost as the result, because the actual receiver could be some subtype of Top.
E.g. in `ref.m(x)` if `ref` has static type Top, it could still point to an A object and we need to use Lost to avoid a confusion.

For constructor invocations, that argument doesn't hold - we always instantiate the object with the most concrete type - there is an actual Top object and invoking it with Top is okay.

This is different from PICO and Universes, where one can never have an object with concrete type `Readonly`.

I don't know what to do with this now... Maybe forbid creating Top objects? Or document that handling of Top is a bit inconsistent?

**Review by wmdietl — COMMENTED** [review]
Thanks!

File: framework/src/test/java/viewpointtest/ViewpointTestQualifierHierarchy.java (hunk 1 above)
Comment (Ao-senXiong) [reply framework/src/test/java/viewpointtest/ViewpointTestQualifierHierarchy.java]: Added link the QualifierHierarchy class.

File: framework/src/test/java/viewpointtest/quals/Lost.java (hunk 2 above)
Comment (Ao-senXiong) [reply framework/src/test/java/viewpointtest/quals/Lost.java:15]: see #871

File: framework/tests/viewpointtest/VarargsConstructor.java (hunk 4 above)
Comment (Ao-senXiong) [reply framework/tests/viewpointtest/VarargsConstructor.java:21]: Forbid creating Top objects sounds good to me. Does it apply all systems that use viewpoint adapation?

File: framework/tests/viewpointtest/LostNonReflexive.java (hunk 3 above)
Comment (Ao-senXiong) [reply framework/tests/viewpointtest/LostNonReflexive.java:10]: I am not sure, let me check...

Comment (Ao-senXiong) [reply framework/tests/viewpointtest/LostNonReflexive.java:10]: I will leave this to new issue #873

File: framework/src/test/java/viewpointtest/quals/Lost.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import org.checkerframework.framework.qual.SubtypeOf;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * {@link Lost} is type annotation that indicates the relationship can not be expressed. It is the
```
Comment (wmdietl) [inline framework/src/test/java/viewpointtest/quals/Lost.java:12]:
```suggestion
 * The {@link Lost} qualifier indicates that a relationship cannot be expressed. It is the
```

Might need to re-flow.

File: framework/src/test/java/viewpointtest/quals/Lost.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * result of viewpoint adaptation that combines {@link Top} and {@link ReceiverDependentQual}.
+ *
+ * <p>It is not reflexive in subtyping relationship and the only subtype for {@link Lost} is {@link
+ * Bottom}.
```
Comment (wmdietl) [inline framework/src/test/java/viewpointtest/quals/Lost.java:16]:
```suggestion
 * <p>It is not reflexive in the subtyping relationship and the only subtype for {@link Lost} is {@link
 * Bottom}.
```

File: framework/tests/viewpointtest/SuperConstructorCalls.java [hunk 7]
```diff
@@ -18,6 +18,7 @@ public Inner() {
         }
 
         public Inner(@Top Object objTop) {
+            // :: error: (argument.type.incompatible)
```
Comment (wmdietl) [inline framework/tests/viewpointtest/SuperConstructorCalls.java:21]: A comment would be helpful here. I assume the constructor is `@Top` by default, which then leads to `@Lost` in the super constructor signature.

File: framework/tests/viewpointtest/VarargsConstructor.java (hunk 4 above)
Comment (wmdietl) [reply framework/tests/viewpointtest/VarargsConstructor.java:21]: Not to all VP systems. But if VP with some qualifier results in Lost, that qualifier probably can't be created - Lost comes from not knowing some relationship. If one can create an object of that type, one knows the relationship.
So in this system, forbidding Top makes sense.

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks for wrapping this up!

File: framework/tests/viewpointtest/SuperConstructorCalls.java (hunk 7 above)
Comment (Ao-senXiong) [reply framework/tests/viewpointtest/SuperConstructorCalls.java:24]: Added!

File: framework/tests/viewpointtest/VarargsConstructor.java (hunk 4 above)
Comment (Ao-senXiong) [reply framework/tests/viewpointtest/VarargsConstructor.java:22]: Done!

File: framework/src/main/java/org/checkerframework/common/basetype/messages.properties [hunk 8]
```diff
@@ -2,6 +2,7 @@
 receiver.invalid=incompatible types.%nfound   : %s%nrequired: %s
 type.anno.before.modifier=write type annotation %s immediately before type, after modifiers %s
 type.anno.before.decl.anno=write type annotations %s immediately before type, after declaration annotation %s
+new.class.type.invalid=the annotations %s do not need be applied in object creations
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/common/basetype/messages.properties:5]: Can you move this further down to other `type.invalid` error messages.
While we're at it, can you improve the grammar?

File: framework/src/test/java/viewpointtest/ViewpointTestAnnotatedTypeFactory.java [hunk 9]
```diff
@@ -3,19 +3,33 @@
...
+
 import viewpointtest.quals.A;
 import viewpointtest.quals.B;
 import viewpointtest.quals.Bottom;
+import viewpointtest.quals.Lost;
 import viewpointtest.quals.PolyVP;
 import viewpointtest.quals.ReceiverDependentQual;
 import viewpointtest.quals.Top;
 
+/** The annotated type factory for the viewpointtest type-system. */
```
Comment (wmdietl) [inline framework/src/test/java/viewpointtest/ViewpointTestAnnotatedTypeFactory.java:22]:
```suggestion
/** The annotated type factory for the Viewpoint Test Checker. */
```

It would be nice to use consistent wording for the different javadocs.

File: framework/src/test/java/viewpointtest/ViewpointTestViewpointAdapter.java [hunk 10]
```diff
@@ -8,12 +8,15 @@
 
 import javax.lang.model.element.AnnotationMirror;
 
+import viewpointtest.quals.Lost;
 import viewpointtest.quals.ReceiverDependentQual;
 import viewpointtest.quals.Top;
 
+/** The viewpoint adapter for the ViewpointTest checker. */
```
Comment (wmdietl) [inline framework/src/test/java/viewpointtest/ViewpointTestViewpointAdapter.java:15]:
```suggestion
/** The viewpoint adapter for the Viewpoint Test Checker. */
```

File: framework/src/test/java/viewpointtest/ViewpointTestVisitor.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
+package viewpointtest;
+
+import com.sun.source.tree.NewClassTree;
+
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.basetype.BaseTypeVisitor;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+
+/** The visitor for the viewpointtest type-system. */
```
Comment (wmdietl) [inline framework/src/test/java/viewpointtest/ViewpointTestVisitor.java:9]:
```suggestion
/** The visitor for the Viewpoint Test Checker. */
```

File: framework/tests/viewpointtest/PolyConstructor.java [hunk 12]
```diff
@@ -3,9 +3,10 @@
 public class PolyConstructor {
 
     static class MyClass {
-        @SuppressWarnings({"inconsistent.constructor.type", "super.invocation.invalid"})
+        // :: error: (super.invocation.invalid) :: warning: (inconsistent.constructor.type)
         @PolyVP MyClass(@PolyVP Object o) {
-            throw new RuntimeException(" * You are filled with DETERMINATION."); // stub
+            // :: warning: (cast.unsafe.constructor.invocation)
+            throw new @A RuntimeException(" * You are filled with DETERMINATION."); // stub
```
Comment (wmdietl) [inline framework/tests/viewpointtest/PolyConstructor.java:9]: What's the point of this change? Should there be multiple versions?

File: framework/tests/viewpointtest/VarargsConstructor.java [hunk 13]
```diff
@@ -47,13 +54,17 @@ void invokeConstructor(@A Object aObj, @B Object bObj, @Top Object topObj) {
...
     void testAnonymousClass(@A Object aObj, @B Object bObj, @Top Object topObj) {
         Object o =
-                new VarargsConstructor("testStr", new Object()) {
+                // :: warning: (cast.unsafe.constructor.invocation)
+                new @A VarargsConstructor("testStr", new @A Object()) {
                     void foo() {
-                        VarargsConstructor a = new VarargsConstructor("testStr", new Object());
+                        VarargsConstructor a =
+                                // :: warning: (cast.unsafe.constructor.invocation)
+                                new @A VarargsConstructor("testStr", new @A Object());
```
Comment (wmdietl) [inline framework/tests/viewpointtest/VarargsConstructor.java:62]: Is this change illustrating something about `Lost`? Should there be versions of this test without an annotation? A comment would help understand this test.

File: framework/src/test/java/viewpointtest/ViewpointTestQualifierHierarchy.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    public ViewpointTestQualifierHierarchy(
+            Collection<Class<? extends Annotation>> qualifierClasses,
+            Elements elements,
+            GenericAnnotatedTypeFactory<?, ?, ?, ?> atypeFactory) {
+        super(qualifierClasses, elements, atypeFactory);
+    }
+
+    @Override
+    public boolean isSubtypeQualifiers(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+        // Lost is not reflexive and the only subtype is Bottom
```
Comment (wmdietl) [inline framework/src/test/java/viewpointtest/ViewpointTestQualifierHierarchy.java:34]:
```suggestion
        // Lost is not reflexive and the only subtype is Bottom.
```

File: framework/src/test/java/viewpointtest/ViewpointTestVisitor.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+     * @param checker the checker to which this visitor belongs
+     */
+    public ViewpointTestVisitor(BaseTypeChecker checker) {
+        super(checker);
+    }
+
+    @Override
+    public Void visitNewClass(NewClassTree tree, Void p) {
+        AnnotatedTypeMirror Type = atypeFactory.getAnnotatedType(tree);
+        if (Type.hasAnnotation(atypeFactory.TOP)) {
```
Comment (wmdietl) [inline framework/src/test/java/viewpointtest/ViewpointTestVisitor.java:23]: Shouldn't `LOST` also be forbidden?

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks for the fixes and the extensive tests!

File: framework/src/test/java/viewpointtest/ViewpointTestVisitor.java (hunk 15 above)
Comment (Ao-senXiong) [reply framework/src/test/java/viewpointtest/ViewpointTestVisitor.java:23]: Yes, how about `poly` and `RD` qualifier. I left out those two as I feel like we need more discussion.

File: framework/tests/viewpointtest/VarargsConstructor.java (hunk 13 above)
Comment (Ao-senXiong) [reply framework/tests/viewpointtest/VarargsConstructor.java:62]: This change is not about lost, I update this test case because I want those invalid invocation because previous we allow `new @Top` constructor invocation.  More test cases for lost is in `LostNonReflexive.java`

File: framework/tests/viewpointtest/PolyConstructor.java (hunk 12 above)
Comment (Ao-senXiong) [reply framework/tests/viewpointtest/PolyConstructor.java]: I expand the test suite in here.

**Review by wmdietl — APPROVED** [review]
Thanks!

File: framework/src/test/java/viewpointtest/ViewpointTestViewpointAdapter.java [hunk 16]
```diff
@@ -22,10 +25,11 @@ public class ViewpointTestViewpointAdapter extends AbstractViewpointAdapter {
      */
     public ViewpointTestViewpointAdapter(AnnotatedTypeFactory atypeFactory) {
         super(atypeFactory);
-        TOP = AnnotationBuilder.fromClass(atypeFactory.getElementUtils(), Top.class);
+        TOP = atypeFactory.TOP;
         RECEIVERDEPENDENTQUAL =
                 AnnotationBuilder.fromClass(
                         atypeFactory.getElementUtils(), ReceiverDependentQual.class);
+        LOST = atypeFactory.LOST;
```
Comment (wmdietl) [inline framework/src/test/java/viewpointtest/ViewpointTestViewpointAdapter.java:32]:
```suggestion
        TOP = ((ViewpointTestAnnotatedTypeFactory) atypeFactory).TOP;
        RECEIVERDEPENDENTQUAL =
                AnnotationBuilder.fromClass(
                        atypeFactory.getElementUtils(), ReceiverDependentQual.class);
        LOST = ((ViewpointTestAnnotatedTypeFactory) atypeFactory).LOST;
```
